lyse/__main__.py

```diff
@@ -152,18 +152,43 @@ def get_screen_geometry():
 
 class WebServer(ZMQServer):
 
+    @inmain_decorator(wait_for_return=True)
     def handler(self, request_data):
         logger.info('WebServer request: %s' % str(request_data))
         if request_data == 'hello':
             return 'hello'
-        elif request_data == 'get dataframe':
+        elif isinstance(request_data, tuple) and request_data[0]=='get dataframe' and len(request_data)==3:
+            _, n_sequences, filter_kwargs = request_data
             # infer_objects() picks fixed datatypes for columns that are compatible with
             # fixed datatypes, dramatically speeding up pickling. It is called here
             # rather than when updating the dataframe as calling it during updating may
             # call it needlessly often, whereas it only needs to be called prior to
             # sending the dataframe to a client requesting it, as we're doing now.
             app.filebox.shots_model.infer_objects()
-            return app.filebox.shots_model.dataframe
+            df = app.filebox.shots_model.dataframe
+            df = _rangeindex_to_multiindex(df, inplace=False)
```

Comment:
So the use of `inplace=False` here makes perfect sense, but does that mean that we have a deepcopy of the dataframe? And thus actually only need to lock this one line with `inmain`, since we're copying anyway?

Or should we be trying to avoid that performance hit and only doing the index conversion server side if `filter_args` is not None? That means `_rangeindex_to_multiindex` would need to detect if it has already run once so it could always be called on the client side.

I'm not quite sure what the best approach here is.

Reply:
I've spent a few hours trying to figure out pandas' memory management but haven't been able to confidently come to a conclusion about this. The topic of "when does panda return a copy vs return a view" is a bit complicated.

Playing around in the interpreter, I tried the following:

```python
df0 = lyse.data()
df0 = df1.set_index('sequence', inplace=False, drop=False)
df1.loc[:, 'run number'] = 12345
```

That last line modifies `df1` but `df0` remains unchanged, so it seems like `set_index()` did make a deep copy. I suspect that it is guaranteed to always make a deep copy, but I'm definitely not 100% sure. I'm also not sure if it is an immediate copy, or a copy-on-write. If it's a copy-on-write then in `WebServer.handler()` the underlying data (ideally) shouldn't be duplicated since the dataframe copy goes out of scope when `handler()` returns, which is before anything edits the underlying data thanks to the `@inmain_decorator`.

We could make lyse only do the conversion server-side if `filter_args` is provided. Fortunately `_rangeindex_to_multiindex()` is already written to detect if it has already been run. That was implemented to make sure that it still runs when using updated clients with outdated servers, but avoid running it twice when updated clients request from updated servers. One small difference though would be that `df.sort_index()` won't be run before `_extract_n_sequences_from_df()` if  `_rangeindex_to_multiindex()` isn't called server-side. That would affect which rows are returned if they are added to lyse out of order, e.g. if a user adds files to it manually using the "+" button. I actually just moved `df.sort_index()` from `data()` to `_rangeindex_to_multiindex()` in 4c944dd for that very reason.

Reply:
Actually,  I took a look at the source code and it seems that `set_index(inplace=False)` [calls `self.copy()`](https://github.com/pandas-dev/pandas/blob/db08276bc116c438d3fdee492026f8223584c477/pandas/core/frame.py#L4553-L4556) which [makes a deep copy](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.copy.html) by default.

There's still a chance that it's a copy-on-write but from what I've read I get them impression that copy-on-write is on some pandas todo/wishlists but isn't implemented. So I'm becoming more confident that it always does do a deep copy immediately.

Reply:
I'm wondering how resource-heavy making the copy actually is though. I have an instance of lyse running at the moment with ~3,000 shots in it. Requesting the dataframe takes ~0.5 seconds, but running `df = lyse.data(); df.memory_usage(deep=True).sum()` shows that it's still only 15 MB. Also, running `df.copy(deep=True)` only takes about 7 ms. That's not bad, and it still wouldn't be terrible with ~10x more shots assuming things scale linearly.

Reply:
Ok, so I wrote something (below the horizontal rule) about more tests (left for posterity) but I don't know whether they are important to do.

If copying is always 100 times faster than transmitting, then there is very little need to optimise for the case where no filtering/slicing is done (which is I guess what my suggestion boils down to). Could you do a brief test to see how `df = lyse.data();` vs `df.copy(deep=True)` scales with more shots and more columns? There will be some overhead for small numbers of shots (meaning `lyse.data` will stay ~constant while `copy` grows), but at some point I would expect them to grow with some constant ratio. If we confirm that ratio is big (like ~100 as you saw for 3000 shots), then we should always do the copy on the server side, and only have that run in the main loop (probably explicitly do `copy(deep=True)` in mainthread and then do an inplace index modification on the copy in the zmqserver thread). The question is whether 3000 shots is "big" or not!

-----
Hmm, interesting.

Could you do a couple of other checks on the performance of `copy(deep=True)`?

1. Can you confirm it does scale approximately linearly with the number of shots?
2. Could you report the copy performance with a df with those 3000 shots in and then with an additional 3000 shots that have had a different analysis script run on them (this makes the dataframe expand columns and fill in NaNs for the new columns, old shots intersection)?

Also, couple of questions about the shots:

1. How many globals are in the shots?
2. How many analysis result columns are in the dataframe?

Reply:
Good questions. I'm still working on answering some of them, but I'll respond with what I have now before I lose track of it.

> 1. How many globals are in the shots?

About 250.

> 2. How many analysis result columns are in the dataframe?

About 25.

> Can you confirm it (`copy(deep=True)`) does scale approximately linearly with the number of shots?

Running the following code:

```python
from time import perf_counter
import matplotlib.pyplot as plt
import numpy as np
import lyse

df = lyse.data()  # Need at least 3000 runs in dataframe for this test.

times = []
sizes = np.linspace(1, 3001, 31)

for size in sizes:
    size = int(size)
    df_subset = df.iloc[:size]
    start = perf_counter()
    df_subset.copy(deep=True)
    end = perf_counter()
    times.append(end-start)

fig = plt.figure()
axes = fig.add_subplot(111)
axes.plot(
    sizes,
    times,
    marker = 'o',
)
axes.set_title('Dataframe Deep Copy Times')
axes.set_xlabel('Number of Rows')
axes.set_ylabel('Duration (s)')
```

Gave the following plot:

![image](https://user-images.githubusercontent.com/4721629/97787979-e9b09500-1b8b-11eb-9fa1-d965cd72daaa.png)

Which looks pretty linear.

Reply:
> 2 Could you report the copy performance with a df with those 3000 shots in and then with an additional 3000 shots that have had a different analysis script run on them (this makes the dataframe expand columns and fill in NaNs for the new columns, old shots intersection)?

I figured it would take a while to set up that exact test, so I did something similar but slightly different. I used the same dataframe with ~3000 shots, but used the "Mark as not done" feature to run a singleshot lyse script on only ~5 of those runs. That produced 10 results which weren't calculated for the other runs so the dataframe had a lot of empty entries. I ran the singleshot routine on some rows around index 10 in the dataframe so that they would be included in most of the subsets. Re-running that code from the last comment gave the plot below:

![image](https://user-images.githubusercontent.com/4721629/97788211-b111bb00-1b8d-11eb-86eb-6a08be4d9f5a.png)

So the results are pretty similar with several extra columns that have a lot of empty entries.

I'll set up a script to time `df = lyse.data()` and `df.copy(deep=True)` as the size of the dataframe changes while I run an M-LOOP optimization. That'll take a while to complete but hopefully I'll have something by the end of the day.

Reply:
> If copying is always 100 times faster than transmitting, then there is very little need to optimise for the case where no filtering/slicing is done (which is I guess what my suggestion boils down to). Could you do a brief test to see how df = lyse.data(); vs df.copy(deep=True) scales with more shots and more columns? 

This only goes up to ~1700 shots but the scaling is pretty clear. It seems that the time for `data()` is always roughly equal to 100 times the time required for `copy()`.

![image](https://user-images.githubusercontent.com/4721629/97790356-73695e00-1b9e-11eb-8276-c8018bf7a464.png)

I'm not sure why the yellow curve jumps at ~500 entries. My best guess is that it's some internal pandas thing about how small vs big dataframes are copied. At any rate, after that point the two curves scale very similarly.

I should also mention that all of these timing tests were done with a copy of lyse that doesn't include the changes in this PR.

Reply:
Ok. Think I'm happy with this. We can revisit if someone complains about performance.